plugins/chain_plugin/include/eosio/chain_plugin/account_query_db.hpp

```diff
@@ -0,0 +1,126 @@
+#pragma once
+#include <eosio/chain/types.hpp>
+#include <eosio/chain/block_state.hpp>
+#include <eosio/chain/trace.hpp>
+
+namespace eosio::chain_apis {
+   /**
+    * This class manages the ephemeral indices and data that provide the `get_accounts_by_authorizers` RPC call
+    * There is no persistence and the indices/caches are recreated when the class is instantiated based on the
+    * current state of the chain.
+    */
+   class account_query_db {
+   public:
+
+      /**
+       * Instantiate a new account query DB from the given chain controller
+       * The caller is expected to manage lifetimes such that this controller reference does not go stale
+       * for the life of the account query DB
+       * @param chain - controller to read data from
+       */
+      account_query_db( const class eosio::chain::controller& chain );
+      ~account_query_db();
+
+      /**
+       * Allow moving the account query DB (including by assignment)
+       */
+      account_query_db(account_query_db&&);
+      account_query_db& operator=(account_query_db&&);
+
+      /**
+       * Add a transaction trace to the account query DB that has been applied to the contoller even though it may
+       * not yet be committed to by a block.
+       *
+       * @param trace
+       */
+      void cache_transaction_trace( const chain::transaction_trace_ptr& trace );
+
+      /**
+       * Add a block to the account query DB, committing all the cached traces it represents and dumping any
+       * uncommitted traces.
+       * @param block
+       */
+      void commit_block(const chain::block_state_ptr& block );
+
+      /**
+       * parameters for the get_accounts_by_authorizers RPC
+       */
+      struct get_accounts_by_authorizers_params{
+         struct account_level {
+            chain::name actor;
+            chain::name permission;
+         };
+         std::vector<account_level> accounts;
```

Comment:
It is a new type to allow the parsing rules for the API (`"actor"` or `{"actor", "permission}`). The alternative would be to do the `from_variant` of the parent data structure by hand and just use the json library to extract an `fc::variant`.  this seemed like the lessor of two hacks.

Reply:
ended up making this a new struct that inherits from `eosio::chain::permission_level` so that I could make new type-based rules but trivially cast to the underlying type.